src/modules/rlm_mschap/rlm_mschap.c

```diff
@@ -1641,29 +1585,132 @@ static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(void *instance, REQUEST *re
 		}
 	}
 	/*
-	 *	... or a Cleartext-Password, which we now transform into an LM-Password
+	 *	If we can't find an LM-Password, try and create one from password
 	 */
 	if (!lm_password) {
+		uint8_t *p;
+
 		if (password) {
 			RDEBUG2("Found Cleartext-Password, hashing to create LM-Password");
 			lm_password = pair_make_config("LM-Password", NULL, T_OP_EQ);
 			if (!lm_password) {
 				RERROR("No memory");
-			} else {
-				uint8_t *p;
-
-				p = talloc_array(lm_password, uint8_t, LM_DIGEST_LENGTH);
-				fr_pair_value_memsteal(lm_password, p);
-				smbdes_lmpwdhash(password->vp_strvalue, p);
+				return false;
 			}
+
+			p = talloc_array(lm_password, uint8_t, LM_DIGEST_LENGTH);
+			fr_pair_value_memsteal(lm_password, p);
+			smbdes_lmpwdhash(password->vp_strvalue, p);
+
 		/*
 		 *	Only complain if we don't have NT-Password
 		 */
-		} else if ((auth_method == AUTH_INTERNAL) && !nt_password) {
+		} else if (inst->method == AUTH_INTERNAL && !nt_password) {
 			RWDEBUG2("No Cleartext-Password configured.  Cannot create LM-Password");
 		}
 	}
 
+	*lmpw = lm_password;
+	return true;
+}
+
+
+/*
+ *	mod_authenticate() - authenticate user based on given
+ *	attributes and configuration.
+ *	We will try to find out password in configuration
+ *	or in configured passwd file.
+ *	If one is found we will check paraneters given by NAS.
+ *
+ *	If PW_SMB_ACCOUNT_CTRL is not set to ACB_PWNOTREQ we must have
+ *	one of:
+ *		PAP:      PW_USER_PASSWORD or
+ *		MS-CHAP:  PW_MSCHAP_CHALLENGE and PW_MSCHAP_RESPONSE or
+ *		MS-CHAP2: PW_MSCHAP_CHALLENGE and PW_MSCHAP2_RESPONSE
```

Comment:
Probably... that's not my comment ;) (pretty much none of it is mine - I just moved stuff around).

I'll take a look if I get a moment.